[core] Prepend "use-client" directive + add docs and examples for using MUI libraries with Next.js App Router

mj12albert opened the pull request:
Closes [37503](https://github.com/mui/material-ui/issues/37503)

This PR adds the "use-client" directive to components/hooks exported from core libraries, and a CI step to check it.

Here's a [basic demo](https://codesandbox.io/p/sandbox/tender-sanderson-vd22sg?file=%2Fapp%2Fpage.tsx%3A18%2C1) with some `@mui/material` components with  (using default imports) working in Next.js 13's app dir, `ThemeProvider` is not used, no (additional) `use client` directives are needed in `page` or `layout` files.

Here's [the same one](https://codesandbox.io/p/sandbox/distracted-feynman-fvqq3z?file=%2Fapp%2Fpage.tsx%3A16%2C1) but using named imports – turns out `'use client'` is needed on all the `index.js|ts` to make this work.

Summary of other changes:

- Added guides on how to set up and use MUI libraries with Next 13
  - Base UI ([preview](https://deploy-preview-37656--material-ui.netlify.app/base/guides/next-js-app-router))
  - Material UI ([preview](https://deploy-preview-37656--material-ui.netlify.app/material-ui/guides/next-js-app-router))
  - Joy UI ([preview](https://deploy-preview-37656--material-ui.netlify.app/joy-ui/guides/next-js-app-router))
- Updated the Material UI/Next 13 example - built out the layout and page, and simplified the project structure to be a bit closer to the output of create-next-app
- Added examples for Joy UI and Base UI with Next 13
- Material UI's umd build is built with rollup which ignores `use client`, but based on the [CDN example](https://github.com/mui/material-ui/tree/master/examples/material-via-cdn) I think it shouldn't matter (there is no server environment)
- The `@mui/icons-material` build script needs to be run and changes committed, but I will do it in a [separate PR](https://github.com/mui/material-ui/pull/37894) (it results in 20k+ changed files and will make the GitHub UI lag like no tomorrow)

To try out the new example repos, replace the `@mui` packages from `package.json` with the latest version built by CI:

```diff
 "dependencies": {
-  "@mui/material": "latest",
+  "@mui/material": "https://pkg.csb.dev/mui/material-ui/commit/0fb89e43/@mui/material",
-  "@mui/base": "latest",
+  "@mui/base": "https://pkg.csb.dev/mui/material-ui/commit/0fb89e43/@mui/base",
-  "@mui/joy": "latest",
+  "@mui/joy": "https://pkg.csb.dev/mui/material-ui/commit/0fb89e43/@mui/joy",
 },
``` 

- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui/material-ui/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

mnajdova reviewed (commented):
This looks like a great start! I would recommend including the `@mui/material-next` and `@mui/system` packages too. Would be interesting to investigate if we have a component or hook that support RSC, so that we can add logic in the script to skip some files.

samuelsycamore commented on docs/data/base/getting-started/quickstart/quickstart.md:
```diff
@@ -2,6 +2,10 @@
 <p class="description">Get started with Base UI, a library of headless ("unstyled") React UI components and low-level hooks.</p>
 
+:::info
+⚗️ If you are using the Next.js App Router, see our [guide](/material-ui/guides/next-js-app-router/#using-base-ui)
```
Rephrasing to make the link more descriptive.

```suggestion
If you're using Next.js 13.4 or later, see our [Next.js App Router guide](/material-ui/guides/next-js-app-router/#using-base-ui).
```

samuelsycamore commented on docs/data/material/guides/next-js-app-router/next-js-app-router.md:
```diff
@@ -0,0 +1,227 @@
+
+## Next.js and React Server Components
+
+Next.js's App Router implements React Server Components, which is a [new feature](https://github.com/reactjs/rfcs/blob/main/text/0227-server-module-conventions.md#changes-since-v1) in React.
```
```suggestion
The Next.js App Router implements React Server Components, a [new feature](https://github.com/reactjs/rfcs/blob/main/text/0227-server-module-conventions.md#changes-since-v1) introduced in React 18.
```

samuelsycamore commented on docs/data/material/guides/next-js-app-router/next-js-app-router.md:
```diff
@@ -0,0 +1,227 @@
+
+Next.js's App Router implements React Server Components, which is a [new feature](https://github.com/reactjs/rfcs/blob/main/text/0227-server-module-conventions.md#changes-since-v1) in React.
+
+To support the App Router, currently all components and hooks from MUI libraries (Material UI, Joy UI, Base UI etc) are designated as client components and exported with the "use client" directive.
```
```suggestion
To support the App Router, all components and hooks from MUI libraries (Material UI, Joy UI, Base UI etc) are designated as Client Components and exported with the `"use client"` directive.
```

samuelsycamore commented on docs/data/material/guides/next-js-app-router/next-js-app-router.md:
```diff
@@ -0,0 +1,227 @@
+To support the App Router, currently all components and hooks from MUI libraries (Material UI, Joy UI, Base UI etc) are designated as client components and exported with the "use client" directive.
+
+:::info
+Don't need a guide? Use one of these example repos and get started:
```
Rather than framing it in terms of need/skill, how about something like this?

```suggestion
Starting fresh on a new App Router-based project?
Jump right into the code with one of these example repos:
```

samuelsycamore commented on docs/data/material/guides/next-js-app-router/next-js-app-router.md:
```diff
@@ -0,0 +1,227 @@
+
+## Using Material UI with the default theme
+
+If you are using the default theme, Material UI components will work out-of-the-box in Next.js routing files such as `layout.js` or `page.js` - which are server components by default.
```
Active voice.

```suggestion
If you're using the default theme, you can add Material UI components to Next.js routing files such as `layout.js` or `page.js` (which are Server Components by default) without any additional configuration, as shown below:
```

samuelsycamore commented on docs/data/material/guides/next-js-app-router/next-js-app-router.md:
```diff
@@ -0,0 +1,227 @@
+
+### Theme Registry
+
+To set up the theme context, create a `ThemeRegistry` component that combines the Emotion CacheProvider, the Material UI Theme Provider and the `useServerInsertedHTML` hook from `next/navigation` as follows:
```
```suggestion
To set up the theme context, create a custom `ThemeRegistry` component that combines the Emotion `CacheProvider`, the Material UI `ThemeProvider`, and the `useServerInsertedHTML` hook from `next/navigation` as follows:
```

samuelsycamore commented on docs/data/material/guides/next-js-app-router/next-js-app-router.md:
````diff
@@ -0,0 +1,227 @@
+}
+```
+
+### CSS Injection Order
````
Sentence case (rather than Title Case) for headlines.

```suggestion
### CSS injection order
```

samuelsycamore commented on docs/data/material/guides/next-js-app-router/next-js-app-router.md:
```diff
@@ -0,0 +1,227 @@
+
+<!-- https://github.com/emotion-js/emotion/issues/3059 -->
+
+By default emotion will inject Material UI styles at the bottom of the HTML `<head>`, which gives it precedence over custom styles (e.g. from CSS modules, Tailwind CSS, or even plain CSS).
```
```suggestion
By default, Emotion injects Material UI styles at the bottom of the HTML `<head>`, which gives them precedence over custom styles (for example, from CSS modules, Tailwind CSS, or plain CSS).
```

samuelsycamore commented on docs/data/material/guides/next-js-app-router/next-js-app-router.md:
```diff
@@ -0,0 +1,227 @@
+
+By default emotion will inject Material UI styles at the bottom of the HTML `<head>`, which gives it precedence over custom styles (e.g. from CSS modules, Tailwind CSS, or even plain CSS).
+
+In regular client-side React, the `prepend: true` option is passed to `createCache` to reverse the injection order, so custom styles can override Material UI styles without using `!important`.
```
It's not "regular" anymore! 🤪🙃

```suggestion
In client-side React, the `prepend: true` option is passed to `createCache` to reverse the injection order, so custom styles can override Material UI styles without using `!important`.
```

samuelsycamore commented on docs/data/material/guides/next-js-app-router/next-js-app-router.md:
```diff
@@ -0,0 +1,227 @@
+
+In regular client-side React, the `prepend: true` option is passed to `createCache` to reverse the injection order, so custom styles can override Material UI styles without using `!important`.
+
+Currently `prepend` does not work reliably with the App Router, but you can work around it by wrapping emotion styles in a CSS `@layer` with a modification to the snippet above:
```
```suggestion
Currently, `prepend` does not work reliably with the App Router, but you can work around it by wrapping Emotion styles in a CSS `@layer` with a modification to the snippet above:
```

samuelsycamore commented on docs/data/material/guides/next-js-app-router/next-js-app-router.md:
```diff
@@ -0,0 +1,227 @@
+
+## Using Joy UI
+
+The set up for Joy UI is the same as [Material UI with a custom theme](#using-material-ui-with-a-custom-theme), except replace Material's `ThemeProvider` with Joy UI's `CssVarsProvider`:
```
```suggestion
The setup for Joy UI is the same as [Material UI with a custom theme](#using-material-ui-with-a-custom-theme), except you must replace Material UI's `ThemeProvider` with Joy UI's `CssVarsProvider`:
```

samuelsycamore commented on docs/data/material/guides/next-js-app-router/next-js-app-router.md:
```diff
@@ -0,0 +1,227 @@
+
+## Using Base UI
+
+Base UI is styling-solution-agnostic, if you are using Emotion you can follow the same steps above.
```
```suggestion
Base UI is style-agnostic; if you're using Emotion, then you can follow the same steps outlined for Material UI.
```

samuelsycamore commented on docs/data/material/guides/next-js-app-router/next-js-app-router.md:
```diff
@@ -0,0 +1,227 @@
+
+Base UI is styling-solution-agnostic, if you are using Emotion you can follow the same steps above.
+
+If you are using Tailwind CSS, see this [example repo](https://github.com/mui/material-ui/blob/master/examples/base-next-app-router-tailwind-ts) for a working demo.
```
```suggestion
If you're using Tailwind CSS, see this [example repo](https://github.com/mui/material-ui/blob/master/examples/base-next-app-router-tailwind-ts) for a working demo.
```

samuelsycamore commented on docs/data/material/guides/next-js-app-router/next-js-app-router.md:
```diff
@@ -0,0 +1,227 @@
+
+### Customizing Base UI with function props
+
+A common customization method in Base UI is to pass a function to slots in `slotProps` in order to apply dynamic props to slots. For example, changing the background color by applying a different class when a Button is disabled:
```
To make this a complete sentence.

```suggestion
A common customization method in Base UI is to pass a function to slots in `slotProps` in order to apply dynamic props to slots. For example, you might want to change the background color by applying a different class when a Button is disabled:
```

samuelsycamore commented on docs/data/material/guides/next-js-app-router/next-js-app-router.md:
````diff
@@ -0,0 +1,227 @@
+}
+```
+
+However, this will not work since function props are [non-serializable](https://nextjs.org/docs/getting-started/react-essentials#passing-props-from-server-to-client-components-serialization).
````
```suggestion
Unfortunately, **this will not work** since function props are [non-serializable](https://nextjs.org/docs/getting-started/react-essentials#passing-props-from-server-to-client-components-serialization).
```

samuelsycamore commented on docs/data/material/guides/next-js-app-router/next-js-app-router.md:
````diff
@@ -0,0 +1,227 @@
+```
+
+However, this will not work since function props are [non-serializable](https://nextjs.org/docs/getting-started/react-essentials#passing-props-from-server-to-client-components-serialization).
+It is recommended that these components be ["moved to the leaves"](https://nextjs.org/docs/getting-started/react-essentials#moving-client-components-to-the-leaves) to avoid this issue and improve overall performance.
````
Who recommends this?

```suggestion
Instead, the Next.js team recommends moving components like these ["to the leaves"](https://nextjs.org/docs/getting-started/react-essentials#moving-client-components-to-the-leaves) to avoid this issue and improve overall performance.
```

siriwatknp commented on docs/data/material/guides/next-js-app-router/next-js-app-router.md:
```diff
@@ -0,0 +1,227 @@
+:::info
+Don't need a guide? Use one of these example repos and get started:
+
+- [Material UI](https://github.com/mui/material-ui/blob/master/examples/material-next-app-router-ts)
+- [Joy UI](https://github.com/mui/material-ui/blob/master/examples/joy-next-app-router-ts)
+- [Base UI with Tailwind CSS](https://github.com/mui/material-ui/blob/master/examples/base-next-app-router-tailwind-ts)
```
Would be great to add additional info about the product because some developers might have seen these names the first time.

e.g.
```md
- [Joy UI](…): Our new in-house library with fresh design.
- [Base UI with Tailwind CSS]: A collection of unstyled components.
```

cc @samuelsycamore

mj12albert replied:
I've updated it to:
```
Material UI - A React UI library that implements Material Design
Joy UI - Our newest customizable UI library designed to spark joy
Base UI – A collection of headless components and hooks
```

mj12albert commented on docs/data/material/guides/next-js-app-router/next-js-app-router.md:
```diff
@@ -0,0 +1,227 @@
+# Next.js App Router
```
I agree it's a better user experience to have a different doc per library, the only content that will be duplicated could be the Joy UI setup part and the introduction

mj12albert commented on examples/base-next-app-router-tailwind-ts/src/app/globals.css:
```diff
@@ -0,0 +1,28 @@
+@tailwind utilities;
+
+@layer components {
+  /* This is an alternative to using function props and ownerState to style a component, the corresponding Switch component can be rendered in page.tsx without a use-client directive, but you cannot easily import the class names like in JS/TS.
+
+  However this may not be a viable pattern since you'll typically pass other non-serializable props like event handers to components, which means you will have to add your own 'use-client' boundaries in your app anyway. */
```
@michaldudak do you think this is worth putting in the docs?

michaldudak replied:
I don't think so. I believe the way to use Base UI components with heavy customization (like using `slotProps`) will be done by wrapping the Base UI component in another custom component. I think it's reasonable to assume this wrapper will also have to be a client component.

mnajdova commented:
I pushed one commit to fix the script on Windows - https://github.com/mui/material-ui/pull/37656/commits/6f122c204086af4fba17538003f17f5a2c1a9a1b

I pushed another commit to change the SVG template file we use - https://github.com/mui/material-ui/pull/37656/commits/b07e22ee19816d338c8f45bf729aee8748033739 For this to take effect we will need to re-generate all the icons - there is a command for it, I will leave it to you.

One last thing, I believe we should include the @mui/utils package as well in the projects on the script

mj12albert commented:
> I pushed one commit to fix the script on Windows

Thanks ~ I thought using `os.EOL` was supposed to do it but couldn't test it myself

mj12albert commented:
@samuelsycamore I've split up the docs and adjusted the content a bit, also added _another_ callout clarifying RSC vs SSR, appreciate another review 🙏

mwskwong commented:
> ```diff
>  "dependencies": {
> -  "@mui/material": "latest",
> +  "@mui/material": "https://pkg.csb.dev/mui/material-ui/commit/3ae05a99/@mui/material",
> -  "@mui/base": "latest",
> +  "@mui/base": "https://pkg.csb.dev/mui/material-ui/commit/3ae05a99/@mui/base",
> -  "@mui/joy": "latest",
> +  "@mui/joy": "https://pkg.csb.dev/mui/material-ui/commit/3ae05a99/@mui/joy",
>  },
> ```

Just tested on this a bit. It seems that it does work on direct import, but not in named import. 

e.g. in the [sandbox](https://codesandbox.io/p/sandbox/old-https-t6ky8q?file=%2Fapp%2Fpage.tsx%3A18%2C1) of the OP, if I update the import from `import Box from "@mui/material/Box";` to `import { Box } from "@mui/material";`, then the `use client` error will appear.

mj12albert commented:
> Just tested on this a bit. It seems that it does work on direct import, but not in named import. 

@mwskwong Thanks for spotting this and for trying it out ~ 

I will fix it!

mwskwong commented:
> > Just tested on this a bit. It seems that it does work on direct import, but not in named import.
> 
> @mwskwong Thanks for spotting this and for trying it out ~
> 
> I will work on a fix

FYR, I think the "issue" applies to other packages as well (e.g. Joy). Not sure whether it applies to all of them though.

michaldudak commented on docs/data/base/guides/next-js-app-router/next-js-app-router.md:
```diff
@@ -0,0 +1,97 @@
+
+The Next.js App Router implements React Server Components, a [new feature](https://github.com/reactjs/rfcs/blob/main/text/0227-server-module-conventions.md#changes-since-v1) introduced in React 18.
+
+To support the App Router, currently all components and hooks from Base UI and other MUI libraries are designated as Client Components and exported with the `"use client"` directive.
```
I wonder if it's correct to say that **hooks** are client components

michaldudak commented on docs/data/base/guides/next-js-app-router/next-js-app-router.md:
```diff
@@ -0,0 +1,97 @@
+
+### Using callbacks for slot props
+
+A common customization method in Base UI is to pass a callback to slots in `slotProps` in order to apply dynamic props to slots. For example, you might want to change the background color by applying a different class when a Button is disabled:
```
Too many "slots" ;)

```suggestion
A common customization method in Base UI is to pass a callback to slots in `slotProps` in order to apply dynamic props. For example, you might want to change the background color by applying a different class when a Button is disabled:
```

michaldudak commented on docs/data/base/guides/next-js-app-router/next-js-app-router.md:
````diff
@@ -0,0 +1,97 @@
+}
+```
+
+Unfortunately, **this will not work** since function props are [non-serializable](https://nextjs.org/docs/getting-started/react-essentials#passing-props-from-server-to-client-components-serialization).
````
Let's underline that it won't work **in a server component** and it's still valid code with `"use client"`.

michaldudak commented on examples/base-next-app-router-tailwind-ts/src/app/components.tsx:
```diff
@@ -0,0 +1,157 @@
+// https://nextjs.org/docs/getting-started/react-essentials#passing-props-from-server-to-client-components-serialization
+
+export function SelectOption({
+  slotProps = {
```
IMO it's more readable to destructure props in the function body. It took me a while to figure out where does the function actually start.

samuelsycamore commented on docs/data/base/guides/next-js-app-router/next-js-app-router.md:
```diff
@@ -0,0 +1,97 @@
+
+### Emotion
+
+If you're using Emotion, or something Emotion-based like MUI System, you can follow the [same steps](/material-ui/guides/next-js-app-router/#using-material-ui-with-a-custom-theme) outlined for Material UI.
```
Maybe we should provide those instructions here in this doc too? Directing readers to the Material UI docs could still lead to confusion. For example, when I visit the page linked, it tells me to import a Material UI component (`ThemeProvider`), but I don't know if that component is also provided by Base UI, or if I have to install Material UI to keep going.

mj12albert replied:
OK – originally I just linked it because I didn't want CSS-in-JS to take up so much space 😓 

I'll simplify the Emotion setup by a little bit and move Tailwind up to be the first section

samuelsycamore commented on docs/data/base/guides/next-js-app-router/next-js-app-router.md:
```diff
@@ -0,0 +1,97 @@
+
+### Tailwind CSS
+
+[Install and initialize](https://tailwindcss.com/docs/guides/nextjs) Tailwind CSS, be sure to add the `app` directory and other directories to `tailwind.config.js`
```
To make the link more informative—readers should know exactly what they'll find if they click the link.

```suggestion
Follow the [Tailwind CSS doc on working with Next.js](https://tailwindcss.com/docs/guides/nextjs), and be sure to add the `app` directory and other directories to `tailwind.config.js`, as shown below:
```

samuelsycamore commented on docs/data/base/guides/next-js-app-router/next-js-app-router.md:
```diff
@@ -0,0 +1,97 @@
+
+Refer to this [example repo](https://github.com/mui/material-ui/blob/master/examples/base-next-app-router-tailwind-ts) for a full working demo of a Next.js 13 app using Base UI and Tailwind CSS.
+
+### Other CSS-in-JS
```
```suggestion
### Other CSS-in-JS libraries
```

samuelsycamore commented on docs/data/base/guides/next-js-app-router/next-js-app-router.md:
```diff
@@ -0,0 +1,97 @@
+
+### Other CSS-in-JS
+
+Base UI works with other CSS-in-JS libraries such as `styled-components`, see the [Next.js docs](https://nextjs.org/docs/app/building-your-application/styling/css-in-js) for more details.
```
```suggestion
To use Next.js with Base UI and styled-components or other CSS-in-JS solutions, follow the [Next.js doc on CSS-in-JS](https://nextjs.org/docs/app/building-your-application/styling/css-in-js).
```

samuelsycamore commented on docs/data/base/guides/next-js-app-router/next-js-app-router.md:
````diff
@@ -0,0 +1,97 @@
+}
+```
+
+Unfortunately, **this will not work in a server component** since function props are [non-serializable](https://nextjs.org/docs/getting-started/react-essentials#passing-props-from-server-to-client-components-serialization).
````
Next.js and React seem to be going with title case for Server Component and Client Component, so we might as well follow the convention.

```suggestion
Unfortunately, **this does not work in a Server Component** since function props are [non-serializable](https://nextjs.org/docs/getting-started/react-essentials#passing-props-from-server-to-client-components-serialization).
```

samuelsycamore commented on docs/data/joy/guides/next-js-app-router/next-js-app-router.md:
```diff
@@ -0,0 +1,153 @@
+
+To support the App Router, currently all components and hooks from Joy UI and other MUI libraries are exported with the `"use client"` directive.
+
+:::warning
```
Same comments as Base doc

samuelsycamore commented on docs/data/joy/guides/next-js-app-router/next-js-app-router.md:
```diff
@@ -0,0 +1,153 @@
+
+## Function props
+
+Props passed from server components - for example `page.js` or other routing files - must be [serializable](https://nextjs.org/docs/getting-started/react-essentials#passing-props-from-server-to-client-components-serialization).
```
em-dashes (—) with no spaces rather than hyphens with spaces.

```suggestion
Props passed from Server Components—for example, `page.js` or other routing files—must be [serializable](https://nextjs.org/docs/getting-started/react-essentials#passing-props-from-server-to-client-components-serialization).
```

samuelsycamore commented on docs/data/joy/guides/next-js-app-router/next-js-app-router.md:
````diff
@@ -0,0 +1,153 @@
+}
+```
+
+However, function props such as event handlers or render props are **non-serializable** - this won't work in a server component:
````
```suggestion
However, function props such as event handlers or render props are **non-serializable**.
```

samuelsycamore commented on docs/data/joy/guides/next-js-app-router/next-js-app-router.md:
````diff
@@ -0,0 +1,153 @@
+```
+
+However, function props such as event handlers or render props are **non-serializable** - this won't work in a server component:
+
+```tsx
+// page.tsx
+import Button from '@mui/joy/Button';
+import Sheet from '@mui/joy/Sheet';
+
+export default function Page() {
+  return (
+    <>
+      <Sheet variant="outlined">
+        {/* Next.js won't render this button without 'use-client' */}
+        <Button
+          variant="outlined"
+          onClick={() => {
+            console.log('handle click');
+          }}
+        >
+          Submit
+        </Button>
+      </Sheet>
+    </>
+  );
+}
+```
+
````
I think we should wrap this whole thing in a big `:::error` callout so it's immediately clear that **this does not work**. So it'd be like:

```suggestion
:::error
This code snippet _doesn't work_, because the Button's click handler is non-serializable:

// code 
:::

samuelsycamore commented on docs/data/material/guides/next-js-app-router/next-js-app-router.md:
```diff
@@ -0,0 +1,233 @@
+Jump right into the code with one of these example repos:
+
+- [Material UI - A React UI library that implements Material Design](https://github.com/mui/material-ui/blob/master/examples/material-next-app-router-ts)
+- [Joy UI - Our newest customizable UI library designed to spark joy](https://github.com/mui/material-ui/blob/master/examples/joy-next-app-router-ts)
+- [Base UI – A collection of headless components and hooks](https://github.com/mui/material-ui/blob/master/examples/base-next-app-router-tailwind-ts)
```
We can drop these links here, since the doc is just about Material UI now.

samuelsycamore commented on docs/data/material/guides/next-js-app-router/next-js-app-router.md:
```diff
@@ -0,0 +1,233 @@
+
+To support the App Router, currently all components and hooks from MUI libraries (Material UI, Joy UI, Base UI etc) are exported with the `"use client"` directive.
+
+:::warning
```
same as Base comments

samuelsycamore commented on docs/data/material/guides/next-js-app-router/next-js-app-router.md:
```diff
@@ -0,0 +1,233 @@
+
+### Theme Registry
+
+To set up the theme context, create a custom `ThemeRegistry` component that combines the Emotion `CacheProvider`, the Material UI `ThemeProvider` and the `useServerInsertedHTML` hook from `next/navigation` as follows:
```
Oxford comma 😇

```suggestion
To set up the theme context, create a custom `ThemeRegistry` component that combines the Emotion `CacheProvider`, the Material UI `ThemeProvider`, and the `useServerInsertedHTML` hook from `next/navigation` as follows:
```

samuelsycamore commented on docs/data/material/guides/next-js-app-router/next-js-app-router.md:
```diff
@@ -0,0 +1,233 @@
+
+<!-- https://github.com/emotion-js/emotion/issues/3059 -->
+
+By default, Emotion inject Material UI styles at the bottom of the HTML `<head>`, which gives them precedence over custom styles – for example if you are customizing Material UI with CSS modules, Tailwind CSS, or even plain CSS.
```
```suggestion
By default, Emotion injects Material UI styles at the bottom of the HTML `<head>`, which gives them precedence over custom styles—for example, if you're customizing Material UI with CSS modules, Tailwind CSS, or even plain CSS.
```

samuelsycamore commented on docs/data/material/guides/next-js-app-router/next-js-app-router.md:
```diff
@@ -0,0 +1,233 @@
+
+## Function props
+
+Props passed from server components - for example `page.js` or other routing files - must be [serializable](https://nextjs.org/docs/getting-started/react-essentials#passing-props-from-server-to-client-components-serialization).
```
```suggestion
Props passed from Server Components—for example, `page.js` or other routing files—must be [serializable](https://nextjs.org/docs/getting-started/react-essentials#passing-props-from-server-to-client-components-serialization).
```

samuelsycamore commented on docs/data/material/guides/next-js-app-router/next-js-app-router.md:
````diff
@@ -0,0 +1,233 @@
+}
+```
+
+However, function props such as event handlers or render props are **non-serializable** - this won't work in a server component:
````
same as Joy comments for this section

samuelsycamore commented on examples/base-next-app-router-tailwind-ts/src/app/globals.css:
```diff
@@ -0,0 +1,28 @@
+@tailwind utilities;
+
+@layer components {
+  /* This is an alternative to using function props and ownerState to style a component, the corresponding Switch component can be rendered in page.tsx without a use-client directive, but you cannot easily import the class names like in JS/TS.
```
```suggestion
  /* This is an alternative to using function props and ownerState to style a component. The corresponding Switch component can be rendered in page.tsx without a use-client directive, but you cannot easily import the class names like in JS/TS.
```

samuelsycamore reviewed (commented):
Docs are getting really close! Great job @mj12albert !

mwskwong commented:
Since `use client` is added to `createSvgIcon`, all icons in `icons-material` can only be used in a client component, or the following error will appear:
![image](https://github.com/mui/material-ui/assets/32725014/b0309c53-d635-4ff3-a017-bc0944013cde)

I wonder is there a better way to handle/workaround it

coder-xiaotian commented:
This SWC plugin can do the same thing.
[https://github.com/coder-xiaotian/swc-useclient](https://github.com/coder-xiaotian/swc-useclient)

mj12albert commented:
> Since `use client` is added to `createSvgIcon`, all icons in `icons-material` can only be used in a client component, or the following error will appear:

This should go away once all the icons are rebuilt – I'm doing that in a separate PR to manage the size of this one

mnajdova reviewed (commented):
The `rsc-build`, the changes and the sandboxes look good 👍 Great work Albert!

samuelsycamore reviewed (approved):
Docs look good to me! Let's get this bad boy out there! 🚀

mj12albert commented on docs/data/joy/guides/next-js-app-router/next-js-app-router.md:
```diff
@@ -0,0 +1,153 @@
+:::warning
+React Server Components does not replace, and is separate from server-side rendering (SSR). Client Components are still server-rendered to HTML.
+
+For more details, see this [explanation](https://github.com/reactwg/server-components/discussions/4) by the React team.
```
I think `this explanation` is enough!

mj12albert commented on docs/data/joy/guides/next-js-app-router/next-js-app-router.md:
```diff
@@ -0,0 +1,153 @@
+
+## Function props
+
+Props passed from server components - for example `page.js` or other routing files - must be [serializable](https://nextjs.org/docs/getting-started/react-essentials#passing-props-from-server-to-client-components-serialization).
```
https://github.com/mui/material-ui/pull/37656#discussion_r1256152087

It should be em dashes with no spaces! @mbrookes

mj12albert commented on docs/data/base/getting-started/quickstart/quickstart.md:
```diff
@@ -2,6 +2,10 @@
 <p class="description">Get started with Base UI, a library of headless ("unstyled") React UI components and low-level hooks.</p>
 
+:::info
+If you're using Next.js 13.4 or later, see our [Next.js App Router guide](/base-ui/guides/next-js-app-router/)
```
Updated to “check out the” – we use "check out" a lot in other parts of the docs as well!

mj12albert commented on docs/data/joy/guides/next-js-app-router/next-js-app-router.md:
```diff
@@ -0,0 +1,153 @@
+
+## Using Joy UI with the App Router
+
+To set up Joy UI, create a custom `ThemeRegistry` component that combines the Emotion `CacheProvider`, Joy UI's `CssVarsProvider` and the `useServerInsertedHTML` hook from `next/navigation` as follows:
```
@mbrookes Couldn't find the comment right now but TLDR – this "Theme Registry" ideally should come from `@emotion`, and they will accept a PR to make it an experimental package which we may do!

siriwatknp commented on docs/data/material/getting-started/installation/installation.md:
```diff
@@ -2,6 +2,14 @@
 <p class="description">Install Material UI, the world's most popular React UI framework.</p>
 
+:::warning
+MUI is currently working on supporting React Server Components in Material UI.
```
Make sense to me.

siriwatknp reviewed (approved):
👍 Great work @mj12albert

oliviertassinari commented on examples/material-next-app-router-ts/src/app/page.tsx:
```diff
@@ -1,6 +1,78 @@
+  return (
+    <Box
+      sx={{
+        display: 'flex',
+      }}
+    >
+      <Box>
+        <Alert severity="info" sx={{ mt: 2, mb: 5 }}>
+          <AlertTitle>Hello 👋</AlertTitle>
+          This app uses the Next.js App Router and Material UI v5.
+        </Alert>
+
+        <Grid container rowSpacing={3} columnSpacing={3}>
+          <Grid xs={6}>
+            <MediaCard
+              heading="CMYK"
+              text="The CMYK color model (also known as process color, or four color) is a subtractive color model, based on the CMY color model, used in color printing, and is also used to describe the printing process itself."
+            />
+          </Grid>
+          <Grid xs={6}>
+            <MediaCard
+              heading="HSL and HSV"
+              text="HSL (for hue, saturation, lightness) and HSV (for hue, saturation, value; also known as HSB, for hue, saturation, brightness) are alternative representations of the RGB color model, designed in the 1970s by computer graphics researchers."
```
I imagine it was helpful to test the whole PR end-to-end but examples are not templates. Could we revert these changes? Thanks

samuelsycamore replied:
It would be nice to circle back to a conversation we had a couple months ago about standardizing the layout for boilerplate examples. cc @danilo-leal

oliviertassinari replied:
Next.js approach makes a lot of sense for me, once the setup works, they link you a template https://vercel.com/templates?framework=next.js.

@samuelsycamore I believe we already have a standard layout, do you mean improving it and enforcing it to all the examples?

samuelsycamore replied:
Yep that's the idea, standardize something a little more neat and polished like the pretty startup pages used by Vercel, Vite, etc.

danilo-leal replied:
Adding to the list ⎯ I think this adds tons of value! 👍 Also, having a global "Templates" page, complementary to the individual ones found in each product's docs, also looks like a good idea!

oliviertassinari replied:
I added the idea to https://www.notion.so/mui-org/example-Simplify-them-to-link-to-templates-546e5a4574ae4a32b0e684ad1c1a54d0
